Source/WebGPU/WGSL/AST/ASTDumper.cpp

```diff
@@ -0,0 +1,403 @@
+/*
+ * Copyright (C) 2022 Apple Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "ASTDumper.h"
+
+#include "AST.h"
+#include <wtf/DataLog.h>
+#include <wtf/EnumTraits.h>
+#include <wtf/SetForScope.h>
+
+namespace WGSL::AST {
+
+struct Indent {
+    Indent(Dumper& dumper)
+        : m_scope(dumper.m_indent, dumper.m_indent + "    ")
+    { }
+    SetForScope<String> m_scope;
+};
+
+static auto bumpIndent(Dumper& dumper) -> Indent
+{
+    return Indent(dumper);
+}
+
+template<typename T, typename J>
+void Dumper::visitVector(T& nodes, J joiner)
+{
+    if (nodes.isEmpty())
+        return;
+
+    visit(nodes[0]);
+    for (size_t n = 1; n < nodes.size(); ++n) {
+        m_out.print(joiner);
+        visit(nodes[n]);
+    }
+}
+
+String Dumper::toString()
+{
+    return m_out.toString();
+}
+
+#define INDENT(...) m_out.print(m_indent __VA_OPT__(,) __VA_ARGS__)
+#define NEWLINE()   m_out.print("\n");
+#define NL_INDENT() m_out.print("\n", m_indent)
+#define NODENT(...) m_out.print(__VA_ARGS__)
+#define SPACE()     m_out.print(" ");
```

Comment:
I think we should not use `#define`s for these

Reply:
So you'd rather see `m_out.print(m_indent, "enable ", directive.name(), ";");` instead of `INDENT("enable ", directive.name(), ";");`?